# **Consumer Loan Portfolio Analysis & Credit Risk Modelling**

## Data Cleaning & Preparation

**Tools:** Python | Pandas | NumPy

This notebook prepares the consumer loan dataset for exploratory analysis, SQL analysis, and predictive modelling.

The cleaning process focuses on validating the dataset structure, standardizing variables, assessing missing values, converting data types, and investigating potential outliers.

---

## Data Loading & Initial Inspection

This section the loading of the raw loan dataset and the initial checks performed to understand its structure and identify potential data quality issues before cleaning.

The dataset was inspected by reviewing its first few observations, shape, data types, missing values, and duplicate records. These checks provided an initial overview of the dataset and helped guide the subsequent cleaning steps.

### Import Libraries

The required Python libraries are imported for data manipulation, numerical analysis, and dataset inspection.

In [1]:
import pandas as pd
import numpy as np

### Load Dataset

The raw consumer loan dataset is loaded into a Pandas DataFrame from the project data directory.

In [2]:
loan_df = pd.read_csv("../data/loan_data.csv")

### Initial Dataset Inspection

The dataset is first inspected to understand its structure, dimensions, data types, and overall completeness before applying any transformations.

In [3]:
loan_df.head()

,id,address_state,application_type,emp_length,emp_title,grade,home_ownership,issue_date,last_credit_pull_date,last_payment_date,...,sub_grade,term,verification_status,annual_income,dti,installment,int_rate,loan_amount,total_acc,total_payment
0,1077430,GA,INDIVIDUAL,< 1 year,Ryder,C,RENT,11-02-2021,13-09-2021,13-04-2021,...,C4,60 months,Source Verified,30000.0,0.0100,59.83,0.1527,2500,4,1009
1,1072053,CA,INDIVIDUAL,9 years,MKC Accounting,E,RENT,01-01-2021,14-12-2021,15-01-2021,...,E1,36 months,Source Verified,48000.0,0.0535,109.43,0.1864,3000,4,3939
2,1069243,CA,INDIVIDUAL,4 years,Chemat Technology Inc,C,RENT,05-01-2021,12-12-2021,09-01-2021,...,C5,36 months,Not Verified,50000.0,0.2088,421.65,0.1596,12000,11,3522
3,1041756,TX,INDIVIDUAL,< 1 year,barnes distribution,B,MORTGAGE,25-02-2021,12-12-2021,12-03-2021,...,B2,60 months,Source Verified,42000.0,0.0540,97.06,0.1065,4500,9,4911
4,1068350,IL,INDIVIDUAL,10+ years,J&J Steel Inc,A,MORTGAGE,01-01-2021,14-12-2021,15-01-2021,...,A1,36 months,Verified,83000.0,0.0231,106.53,0.0603,3500,28,3835


In [4]:
loan_df.shape

(38576, 24)

In [5]:
loan_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38576 entries, 0 to 38575
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   id                     38576 non-null  int64  
 1   address_state          38576 non-null  object 
 2   application_type       38576 non-null  object 
 3   emp_length             38576 non-null  object 
 4   emp_title              37138 non-null  object 
 5   grade                  38576 non-null  object 
 6   home_ownership         38576 non-null  object 
 7   issue_date             38576 non-null  object 
 8   last_credit_pull_date  38576 non-null  object 
 9   last_payment_date      38576 non-null  object 
 10  loan_status            38576 non-null  object 
 11  next_payment_date      38576 non-null  object 
 12  member_id              38576 non-null  int64  
 13  purpose                38576 non-null  object 
 14  sub_grade              38576 non-null  object 
 15  te

In [6]:
loan_df.isna().sum().sort_values(ascending=False)

emp_title                1438
id                          0
application_type            0
address_state               0
emp_length                  0
grade                       0
home_ownership              0
issue_date                  0
last_credit_pull_date       0
last_payment_date           0
loan_status                 0
next_payment_date           0
member_id                   0
purpose                     0
sub_grade                   0
term                        0
verification_status         0
annual_income               0
dti                         0
installment                 0
int_rate                    0
loan_amount                 0
total_acc                   0
total_payment               0
dtype: int64

In [7]:
loan_df.duplicated().sum()

np.int64(0)

## Data Quality Assessment

This section assesses the dataset for potential data quality issues, including invalid date formats, inconsistent categorical values, missing observations, unusual numerical values, and identifier integrity.

### Date Variable Conversion

The dataset contains several date variables stored as text. These variables are converted to datetime format to ensure consistent handling and enable temporal analysis in later stages of the project.

In [8]:
date_columns = [
    "issue_date",
    "last_credit_pull_date",
    "last_payment_date",
    "next_payment_date"
]

In [9]:
loan_df[date_columns].head()

,issue_date,last_credit_pull_date,last_payment_date,next_payment_date
0,11-02-2021,13-09-2021,13-04-2021,13-05-2021
1,01-01-2021,14-12-2021,15-01-2021,15-02-2021
2,05-01-2021,12-12-2021,09-01-2021,09-02-2021
3,25-02-2021,12-12-2021,12-03-2021,12-04-2021
4,01-01-2021,14-12-2021,15-01-2021,15-02-2021


In [10]:
loan_df[date_columns].dtypes

issue_date               object
last_credit_pull_date    object
last_payment_date        object
next_payment_date        object
dtype: object

In [11]:
loan_df[date_columns].nunique()

issue_date                65
last_credit_pull_date    107
last_payment_date        102
next_payment_date        102
dtype: int64

In [12]:
for col in date_columns:
    print(f"\n{col}:")
    print(loan_df[col].dropna().unique()[:10])


issue_date:
['11-02-2021' '01-01-2021' '05-01-2021' '25-02-2021' '17-07-2021'
 '19-11-2021' '11-06-2021' '02-09-2021' '09-02-2021' '22-07-2021']

last_credit_pull_date:
['13-09-2021' '14-12-2021' '12-12-2021' '16-03-2021' '14-06-2021'
 '14-07-2021' '15-06-2021' '13-03-2021' '13-10-2021' '14-02-2021']

last_payment_date:
['13-04-2021' '15-01-2021' '09-01-2021' '12-03-2021' '12-08-2021'
 '13-12-2021' '14-07-2021' '12-10-2021' '16-03-2021' '13-08-2021']

next_payment_date:
['13-05-2021' '15-02-2021' '09-02-2021' '12-04-2021' '12-09-2021'
 '13-01-2022' '14-08-2021' '12-11-2021' '16-04-2021' '13-09-2021']


In [13]:
for col in date_columns:
    loan_df[col] = pd.to_datetime(
        loan_df[col],
        format="%d-%m-%Y"
    )

In [14]:
loan_df[date_columns].dtypes

issue_date               datetime64[ns]
last_credit_pull_date    datetime64[ns]
last_payment_date        datetime64[ns]
next_payment_date        datetime64[ns]
dtype: object

In [15]:
loan_df[date_columns].head()

,issue_date,last_credit_pull_date,last_payment_date,next_payment_date
0,2021-02-11,2021-09-13,2021-04-13,2021-05-13
1,2021-01-01,2021-12-14,2021-01-15,2021-02-15
2,2021-01-05,2021-12-12,2021-01-09,2021-02-09
3,2021-02-25,2021-12-12,2021-03-12,2021-04-12
4,2021-01-01,2021-12-14,2021-01-15,2021-02-15


The date variables have been successfully converted from string format to `datetime64[ns]`, providing a consistent format for subsequent analysis.

### Categorical Variable Inspection

Categorical variables are reviewed to identify the number of unique categories and check for inconsistent or unexpected values.

Variables with a small number of categories are examined through their value distributions to confirm that their labels are consistent and suitable for analysis.

In [16]:
categorical_columns = loan_df.select_dtypes(include="object").columns
categorical_columns

Index(['address_state', 'application_type', 'emp_length', 'emp_title', 'grade',
       'home_ownership', 'loan_status', 'purpose', 'sub_grade', 'term',
       'verification_status'],
      dtype='object')

In [17]:
loan_df[categorical_columns].nunique().sort_values()

application_type           1
term                       2
verification_status        3
loan_status                3
home_ownership             5
grade                      7
emp_length                11
purpose                   14
sub_grade                 35
address_state             50
emp_title              28525
dtype: int64

In [18]:
for col in categorical_columns:
    if loan_df[col].nunique() <= 20:
        print(f"\n{col}:")
        print(loan_df[col].value_counts(dropna=False))


application_type:
application_type
INDIVIDUAL    38576
Name: count, dtype: int64

emp_length:
emp_length
10+ years    8870
< 1 year     4575
2 years      4382
3 years      4088
4 years      3428
5 years      3273
1 year       3229
6 years      2228
7 years      1772
8 years      1476
9 years      1255
Name: count, dtype: int64

grade:
grade
B    11674
A     9689
C     7904
D     5182
E     2786
F     1028
G      313
Name: count, dtype: int64

home_ownership:
home_ownership
RENT        18439
MORTGAGE    17198
OWN          2838
OTHER          98
NONE            3
Name: count, dtype: int64

loan_status:
loan_status
Fully Paid     32145
Charged Off     5333
Current         1098
Name: count, dtype: int64

purpose:
purpose
Debt consolidation    18214
credit card            4998
other                  3824
home improvement       2876
major purchase         2110
small business         1776
car                    1497
wedding                 928
medical                 667
moving              

### Remove Constant Variable

`application_type` contains only the `INDIVIDUAL` category across all observations. Since the variable provides no variation or discriminatory information, it is removed from the analytical dataset.

In [19]:
loan_df.drop(columns="application_type", inplace=True)

In [20]:
loan_df.shape

(38576, 23)

### Missing value assessment

The `emp_title` variable contains 1,438 missing observations, representing approximately 3.7% of the dataset. It also contains a very high number of unique values, making it difficult to standardize and use effectively in the analysis.

Given its missingness and high cardinality, `emp_title` is removed from the dataset rather than imputing or grouping the values arbitrarily.

In [21]:
missing = loan_df.isna().sum().sort_values(ascending=False)
missing = missing[missing > 0]

missing

emp_title    1438
dtype: int64

In [22]:
missing_pct = (loan_df.isna().mean() * 100).sort_values(ascending=False)
missing_pct = missing_pct[missing_pct > 0]

missing_pct

emp_title    3.727706
dtype: float64

In [23]:
loan_df.drop(columns="emp_title", inplace=True)

In [24]:
loan_df.isnull().sum()

id                       0
address_state            0
emp_length               0
grade                    0
home_ownership           0
issue_date               0
last_credit_pull_date    0
last_payment_date        0
loan_status              0
next_payment_date        0
member_id                0
purpose                  0
sub_grade                0
term                     0
verification_status      0
annual_income            0
dti                      0
installment              0
int_rate                 0
loan_amount              0
total_acc                0
total_payment            0
dtype: int64

### Standardize Categorical Variables

Categorical values are standardized to maintain consistent formatting and prevent equivalent categories from being treated as separate values because of differences in whitespace or representation.

In [25]:
loan_df.columns = (
    loan_df.columns
    .str.strip()
    .str.lower()
    .str.replace(" ", "_")
)

In [26]:
categorical_columns = loan_df.select_dtypes(include="object").columns

for col in categorical_columns:
    loan_df[col] = loan_df[col].str.strip()

### Convert Loan Term

The `term` variable contains the loan duration as text, such as `36 months` and `60 months`.

The numerical duration is extracted and stored as `term_months` so that loan tenure can be analysed as a numerical variable.

In [27]:
loan_df["term"] = loan_df["term"].str.replace(" months", "").astype(int)
loan_df.rename(columns={"term": "term_months"}, inplace=True)

### Convert Employment Length

`emp_length` represents employment tenure using text categories ranging from less than one year to ten or more years.

The categories are converted into numerical years to make employment tenure suitable for quantitative analysis and modelling.

In [28]:
emp_length_mapping = {
    "< 1 year": 0,
    "1 year": 1,
    "2 years": 2,
    "3 years": 3,
    "4 years": 4,
    "5 years": 5,
    "6 years": 6,
    "7 years": 7,
    "8 years": 8,
    "9 years": 9,
    "10+ years": 10
}

loan_df["emp_length"] = loan_df["emp_length"].map(emp_length_mapping)

In [29]:
print(loan_df["term_months"].unique())
print(loan_df["emp_length"].unique())

[60 36]
[ 0  9  4 10  3  5  1  6  2  7  8]


### Numerical Variable Assessment

Numerical variables are assessed to understand their ranges, distributions, and potential extreme observations before proceeding with further analysis.

The assessment includes:

- Reviewing the minimum and maximum values of the numerical variables.
- Calculating skewness to assess the shape and symmetry of the distributions.
- Examining selected percentiles to understand the spread and identify potential extreme values.
- Investigating observations with unusually high or low values alongside relevant loan characteristics.
- Assessing whether extreme observations provide sufficient evidence of potential data-quality issues.

Extreme values are not removed automatically. Any decision to retain or remove observations is based on further examination of the data rather than magnitude alone.

#### Examine Minimum and Maximum Values

Minimum and maximum values are examined to understand the range of each numerical variable (excluding `id` and `member_id`) and identify potentially unusual observations.

In [30]:
numerical_columns = loan_df.select_dtypes(include=np.number).columns.tolist()

numerical_columns.remove("id")
numerical_columns.remove("member_id")

In [31]:
loan_df[numerical_columns].min()

emp_length          0.0000
term_months        36.0000
annual_income    4000.0000
dti                 0.0000
installment        15.6900
int_rate            0.0542
loan_amount       500.0000
total_acc           2.0000
total_payment      34.0000
dtype: float64

In [32]:
loan_df[numerical_columns].max()

emp_length       1.000000e+01
term_months      6.000000e+01
annual_income    6.000000e+06
dti              2.999000e-01
installment      1.305190e+03
int_rate         2.459000e-01
loan_amount      3.500000e+04
total_acc        9.000000e+01
total_payment    5.856400e+04
dtype: float64

#### Assess Distribution and Skewness

Skewness is calculated to assess the symmetry of the numerical variables.

The results show substantial positive skewness in `annual_income`, while variables such as `dti` are comparatively symmetric.

In [33]:
loan_df[numerical_columns].skew()

emp_length        0.200245
term_months       1.047545
annual_income    31.074172
dti              -0.029921
installment       1.120447
int_rate          0.292264
loan_amount       1.050775
total_acc         0.823141
total_payment     1.333828
dtype: float64

#### Examine Percentile Distributions

Selected percentiles were examined to understand the spread of the numerical variables and identify potential extreme values.

`annual_income` shows the most pronounced upper-tail behavior. While 99% of observations have an income below $235,000, the maximum reaches $6,000,000, which is consistent with its high positive skewness.

`loan_amount`, `installment`, `total_acc`, and `total_payment` also show right-skewed distributions, but their extreme values remain within plausible ranges.

At this stage, extreme observations are not removed automatically. Their validity will be assessed further during exploratory data analysis.

In [34]:
loan_df[numerical_columns].quantile([0.01, 0.25, 0.50, 0.75, 0.95, 0.99])

,emp_length,term_months,annual_income,dti,installment,int_rate,loan_amount,total_acc,total_payment
0.01,0.0,36.0,15000.0,0.003600,38.9300,0.0542,1200.0,4.0,881.00
0.25,2.0,36.0,41500.0,0.082100,168.4500,0.0932,5500.0,14.0,5633.00
0.50,4.0,36.0,60000.0,0.134200,283.0450,0.1186,10000.0,20.0,10042.00
0.75,9.0,60.0,83200.5,0.185900,434.4425,0.1459,15000.0,29.0,16658.00
0.95,10.0,60.0,144000.0,0.238300,767.6725,0.1862,25000.0,43.0,30320.25
0.99,10.0,60.0,235000.0,0.266625,932.1700,0.2099,35000.0,56.0,42347.50


#### Investigate Extreme Observations

Variables with unusually high skewness are investigated further by examining their extreme observations alongside relevant loan characteristics. Extreme values are not automatically removed because a large value is not necessarily an erroneous value.

The extreme observations are primarily concentrated in `annual_income`, with several borrowers reporting incomes substantially above the typical range. However, these observations are associated with plausible loan amounts, grades, DTI values, and loan outcomes, suggesting that they should not be removed solely because of their magnitude.

For example, the borrower with an annual income of $6,000,000 has a $5,000 loan and a DTI of 0.01, and the loan was fully paid. Other high-income observations also show varying loan characteristics and outcomes.

Therefore, these observations are retained at this stage. Further analysis during EDA will determine whether they represent legitimate high-income borrowers or potential data-quality issues.

In [35]:
loan_df.nlargest(10, "annual_income")[["annual_income", "loan_amount", "dti", "grade", "loan_status"]]

,annual_income,loan_amount,dti,grade,loan_status
26333,6000000.0,5000,0.0001,C,Fully Paid
27136,3900000.0,25000,0.0020,B,Fully Paid
7849,2039784.0,8450,0.0013,C,Fully Paid
3844,1900000.0,1500,0.0057,A,Fully Paid
11725,1782000.0,12025,0.0131,C,Fully Paid
28465,1440000.0,6400,0.0000,A,Fully Paid
8557,1362000.0,6600,0.0062,A,Fully Paid
21769,1250000.0,15775,0.2181,C,Charged Off
8482,1200000.0,12000,0.0917,A,Fully Paid
17718,1200000.0,10000,0.0146,A,Fully Paid


**Insight**

The numerical variables were assessed using descriptive statistics, skewness, and percentile distributions to identify potential data-quality issues and extreme observations.

`annual_income` showed substantial right skewness (31.07), with a maximum value of $6,000,000 compared with a median of $60,000. The extreme observations were investigated alongside loan amount, DTI, grade, and loan status. The values appeared plausible and did not provide sufficient evidence of data-entry errors.

The remaining numerical variables showed moderate or low skewness and plausible minimum and maximum values.

**Decision:** No numerical observations were removed or capped at this stage. Extreme values in `annual_income` were retained as potentially legitimate borrower characteristics. Any transformation required to address skewness will be considered during exploratory analysis and/or model preparation.

### Remove Identifier Variables

`id` and `member_id` function as unique identifiers rather than borrower or loan characteristics.

They are removed from the analytical dataset because they do not provide meaningful predictive information for credit-risk analysis and are not required for the modelling workflow.

In [36]:
loan_df[["id", "member_id"]].nunique()

id           38576
member_id    38576
dtype: int64

In [37]:
loan_df[["id"]].duplicated().sum(), loan_df[["member_id"]].duplicated().sum()

(np.int64(0), np.int64(0))

In [38]:
loan_df[["id", "member_id"]].isna().sum()

id           0
member_id    0
dtype: int64

In [39]:
loan_df.drop(columns=["id", "member_id"], inplace=True)

## Final Data Validation

The cleaned dataset was rechecked by reviewing its first few observations, shape, data types, missing values, and duplicate records. These checks confirm that the intended cleaning steps were successfully applied and that the dataset is ready for exploratory data analysis.

In [40]:
loan_df.head()

,address_state,emp_length,grade,home_ownership,issue_date,last_credit_pull_date,last_payment_date,loan_status,next_payment_date,purpose,sub_grade,term_months,verification_status,annual_income,dti,installment,int_rate,loan_amount,total_acc,total_payment
0,GA,0,C,RENT,2021-02-11,2021-09-13,2021-04-13,Charged Off,2021-05-13,car,C4,60,Source Verified,30000.0,0.0100,59.83,0.1527,2500,4,1009
1,CA,9,E,RENT,2021-01-01,2021-12-14,2021-01-15,Fully Paid,2021-02-15,car,E1,36,Source Verified,48000.0,0.0535,109.43,0.1864,3000,4,3939
2,CA,4,C,RENT,2021-01-05,2021-12-12,2021-01-09,Charged Off,2021-02-09,car,C5,36,Not Verified,50000.0,0.2088,421.65,0.1596,12000,11,3522
3,TX,0,B,MORTGAGE,2021-02-25,2021-12-12,2021-03-12,Fully Paid,2021-04-12,car,B2,60,Source Verified,42000.0,0.0540,97.06,0.1065,4500,9,4911
4,IL,10,A,MORTGAGE,2021-01-01,2021-12-14,2021-01-15,Fully Paid,2021-02-15,car,A1,36,Verified,83000.0,0.0231,106.53,0.0603,3500,28,3835


In [41]:
loan_df.shape

(38576, 20)

In [42]:
loan_df.isnull().sum()

address_state            0
emp_length               0
grade                    0
home_ownership           0
issue_date               0
last_credit_pull_date    0
last_payment_date        0
loan_status              0
next_payment_date        0
purpose                  0
sub_grade                0
term_months              0
verification_status      0
annual_income            0
dti                      0
installment              0
int_rate                 0
loan_amount              0
total_acc                0
total_payment            0
dtype: int64

In [43]:
loan_df.duplicated().sum()

np.int64(0)

In [44]:
loan_df.dtypes

address_state                    object
emp_length                        int64
grade                            object
home_ownership                   object
issue_date               datetime64[ns]
last_credit_pull_date    datetime64[ns]
last_payment_date        datetime64[ns]
loan_status                      object
next_payment_date        datetime64[ns]
purpose                          object
sub_grade                        object
term_months                       int64
verification_status              object
annual_income                   float64
dti                             float64
installment                     float64
int_rate                        float64
loan_amount                       int64
total_acc                         int64
total_payment                     int64
dtype: object

### Export Cleaned Dataset

The cleaned dataset was exported as a CSV file for use in the subsequent exploratory data analysis and modeling stages.

In [45]:
loan_df.to_csv("../data/loan_data_cleaned.csv", index=False)

## Data Cleaning Summary

The consumer loan dataset has been cleaned and prepared for exploratory analysis and downstream analytical work, including SQL analysis.

The cleaning process included:

- Loading and performing initial quality checks on the raw dataset
- Reviewing and standardizing categorical variables for consistency
- Removing the constant `application_type` variable
- Assessing missing values and high cardinality in `emp_title` and removing the variable due to its limited analytical usefulness
- Converting date variables to `datetime` format
- Converting `term` into numerical months and `emp_length` into numerical years
- Assessing numerical distributions using descriptive statistics, skewness, and selected percentiles
- Investigating extreme observations, particularly unusually high `annual_income` values, without automatically removing them
- Removing the `id` and `member_id` identifier variables
- Performing final validation checks on the cleaned dataset, including its shape, data types, missing values, duplicate records, and sample observations

The final cleaned dataset contains 20 columns and 38,576 observations with no missing values or duplicate records.

This cleaned dataset provides the foundation for exploratory data analysis. The EDA stage will determine the appropriate dataset to use for predictive modelling based on the variables, target definition, and analytical findings.

The resulting dataset is ready for the next stage of the project: **exploratory data analysis and portfolio risk analysis**.